src/Avalonia.Base/Visual.cs

```diff
@@ -409,7 +409,10 @@ protected static void AffectsRender<T>(params AvaloniaProperty[] properties)
                 {
                     if (e.Sender is T sender)
                     {
-                        sender.InvalidateVisual();
+                        if (sender.IsVisible || e.Property == IsVisibleProperty)
+                        {
+                            sender.InvalidateVisual();
+                        }
```

Comment:
In many controls we call InvalidateVisual directly from OnPropertyChanged, avoiding AffectsRender